### PR TITLE
Errkit migration 1 (pkg/kube)

### PR DESCRIPTION
**Pull request by e-sumin**
## Change Overview

In accordance with decision to migrate to our own errors package (look here https://github.com/kanisterio/kanister/issues/1838), this PR contains migration of `pkg/kube` to errkit package. Further migration PRs are expected.

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/kube/exec.go [hunk 1]
```diff
@@ -113,32 +120,24 @@ func ExecOutput(cli kubernetes.Interface, namespace, pod, container string, comm
...
-	errbuf := &bytes.Buffer{}
-	if options.Stderr == nil {
-		options.Stderr = errbuf
+		return err
 	}
 
 	errCh := execStream(kubeCli, config, options)
 	err = <-errCh
-	return strings.TrimSpace(outbuf.String()), strings.TrimSpace(errbuf.String()), errors.Wrap(err, "Failed to exec command in pod")
+	if err != nil {
```
Comment (hairyhum) [inline pkg/kube/exec.go:136]: This can potentially be a part of errkit.Wrap

File: pkg/kube/pod_controller.go [hunk 2]
```diff
@@ -29,10 +29,10 @@ import (
 )
 
 var (
-	ErrPodControllerNotInitialized    = errors.New("pod has not been initialized")
-	ErrPodControllerPodAlreadyStarted = errors.New("pod has already been started")
-	ErrPodControllerPodNotReady       = errors.New("pod is not yet ready")
-	ErrPodControllerPodNotStarted     = errors.New("pod is not yet started")
+	ErrPodControllerNotInitialized    = errkit.NewPureError("pod has not been initialized")
```
Comment (hairyhum) [inline pkg/kube/pod_controller.go:32]: Is that still in? I can't find it in the errkit code.

File: pkg/kube/pod.go [hunk 3]
```diff
@@ -412,16 +415,20 @@ func WaitForPodCompletion(ctx context.Context, cli kubernetes.Interface, namespa
 		}
 		containerForLogs = p.Spec.Containers[0].Name
 		if p.Status.Phase == v1.PodFailed {
-			return false, errors.Errorf("Pod %s failed. Pod status: %s", name, p.Status.String())
+			return false, errkit.New("Pod failed", "podName", name, "status", p.Status.String())
 		}
 		return p.Status.Phase == v1.PodSucceeded, nil
 	})
 
+	if err == nil {
```
Comment (hairyhum) [inline pkg/kube/pod.go:423]: I think it would look better as `err != nil` block. But that may be just me.

File: pkg/kube/workload.go [hunk 4]
```diff
@@ -278,7 +278,7 @@ func WaitOnDeploymentReady(ctx context.Context, kubeCli kubernetes.Interface, na
 func WaitOnDeploymentConfigReady(ctx context.Context, osCli osversioned.Interface, kubeCli kubernetes.Interface, namespace, name string) error {
 	return poll.Wait(ctx, func(ctx context.Context) (bool, error) {
 		ok, err := DeploymentConfigReady(ctx, osCli, kubeCli, namespace, name)
-		if apierrors.IsNotFound(errors.Cause(err)) {
+		if apierrors.IsNotFound(errkit.Unwrap(err)) {
```
Comment (hairyhum) [inline pkg/kube/workload.go:281]: Not critical, but I wonder if we can unwrap an error to get a particular type. Would `errors.As` do that?

**Review by hairyhum — APPROVED** [review]

File: pkg/errorchecker/errorchecker.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
+package errorchecker
```
Comment (PrasadG193) [inline pkg/errorchecker/errorchecker.go:1]: Please add license header to maintain consistency with other files

File: pkg/errorchecker/errorchecker.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+var ErrorMessageMatcher check.Checker = errorMatchesChecker{
+	&check.CheckerInfo{Name: "ErrorMatches", Params: []string{"value", "regex"}},
+}
+
+func (checker errorMatchesChecker) Check(
+	params []interface{},
+	names []string,
+) (result bool, errStr string) {
+	if params[0] == nil {
```
Comment (PrasadG193) [inline pkg/errorchecker/errorchecker.go:22]: Would this panic if params is nil? Should we add an additional check to check its not nil?

File: pkg/errorchecker/errorchecker.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+type errorMatchesChecker struct {
+	*check.CheckerInfo
+}
+
+var ErrorMessageMatcher check.Checker = errorMatchesChecker{
+	&check.CheckerInfo{Name: "ErrorMatches", Params: []string{"value", "regex"}},
+}
+
+func (checker errorMatchesChecker) Check(
```
Comment (PrasadG193) [inline pkg/errorchecker/errorchecker.go:18]: Let's add unit tests for this function

Comment (PrasadG193) [inline pkg/errorchecker/errorchecker.go:18]: Please add a comment on the exported functions

File: pkg/kube/pod_controller_test.go [hunk 8]
```diff
@@ -131,7 +130,8 @@ func (s *PodControllerTestSuite) TestPodControllerWaitPod(c *C) {
 			c.Assert(pcp.InWaitForPodReadyPodName, Equals, podControllerPodName)
 			c.Assert(pcp.InWaitForPodReadyNamespace, Equals, podControllerNS)
 			c.Assert(errors.Is(err, pcp.WaitForPodReadyErr), Equals, true)
-			c.Assert(err.Error(), Equals, fmt.Sprintf("Pod failed to become ready in time: %s", simulatedError.Error()))
+
+			c.Assert(err.Error(), Equals, "Pod failed to become ready in time: SimulatedError")
```
Comment (PrasadG193) [inline pkg/kube/pod_controller_test.go:134]: Should we add additional details `simulatedError.Error()` in the message or field?

File: pkg/kube/workload.go (hunk 4 above)
Comment (e-sumin) [reply pkg/kube/workload.go:281]: In fact, under the hood, `IsNotFound` method is doing something like that. 

They don't have `NotFound` error type, instead they have `APIStatus` type. They are doing `errors.As` to this type and then they are checking `reason` and `code`.

File: pkg/errorchecker/errorchecker.go (hunk 6 above)
Comment (e-sumin) [reply pkg/errorchecker/errorchecker.go:37]: `params` can't be `nil` it is passed from `check` framework.
https://github.com/kastenhq/check/blob/v1/helpers.go#L197

So, I suppose, we can rely on that.

File: pkg/errorchecker/errorchecker.go (hunk 7 above)
Comment (e-sumin) [reply pkg/errorchecker/errorchecker.go:33]: Done.

Comment (e-sumin) [reply pkg/errorchecker/errorchecker.go:34]: Done

File: pkg/kube/pod_controller_test.go (hunk 8 above)
Comment (e-sumin) [reply pkg/kube/pod_controller_test.go]: Parameters are not propagated to the error text. However, I've modified the test to simulate a wrapped error now.

**Pull request merged**